### PR TITLE
Feat/provider apis

### PR DESCRIPTION
**Pull request by vibhanshu-ob**
Providers API
```sh
curl --location 'http://127.0.0.1:4000/raga/internal/providers' 
```
Response:
```json
{
    "success": true,
    "data": [
        "openai",
        "text-completion-openai",
        "azure",
        "azure_ai",
        "anthropic",
        "mistral",
        "deepseek",
        "codestral",
        "text-completion-codestral",
        "groq",
        "cerebras",
        "friendliai",
        "vertex_ai-text-models",
        "vertex_ai-chat-models",
        "vertex_ai-code-text-models",
        "vertex_ai-code-chat-models",
        "vertex_ai-language-models",
        "vertex_ai-vision-models",
        "vertex_ai-anthropic_models",
        "vertex_ai-llama_models",
        "vertex_ai-mistral_models",
        "vertex_ai-ai21_models",
        "vertex_ai-image-models",
        "vertex_ai-embedding-models",
        "palm",
        "gemini",
        "cohere_chat",
        "cohere",
        "replicate",
        "openrouter",
        "ai21",
        "nlp_cloud",
        "aleph_alpha",
        "bedrock",
        "sagemaker",
        "together_ai",
        "ollama",
        "deepinfra",
        "perplexity",
        "fireworks_ai",
        "fireworks_ai-embedding-models",
        "anyscale",
        "cloudflare",
        "voyage",
        "databricks"
    ],
    "message": null
}
```

Keys API
```sh
curl --location 'http://127.0.0.1:4000/raga/internal/providers/keys' \
--header 'Content-Type: application/json' \
--data '{
    "provider": "azure"
}'
```
Response:
```json
{
    "success": true,
    "data": [
        "AZURE_API_BASE",
        "AZURE_API_VERSION",
        "AZURE_API_KEY"
    ],
    "message": null
}
```

Models API
```sh
curl --location 'http://127.0.0.1:4000/raga/internal/providers/models' \
--header 'Content-Type: application/json' \
--data '{
    "provider": "openai"
}'
```
Response:
```json
{
    "success": true,
    "data": [
        "gpt-4",
        "gpt-4o",
        "gpt-4o-audio-preview",
        "gpt-4o-audio-preview-2024-10-01",
        "gpt-4o-mini",
        "gpt-4o-mini-2024-07-18",
        "o1-mini",
        "o1-mini-2024-09-12",
        "o1-preview",
        "o1-preview-2024-09-12",
        "chatgpt-4o-latest",
        "gpt-4o-2024-05-13",
        "gpt-4o-2024-08-06",
        "gpt-4-turbo-preview",
        "gpt-4-0314",
        "gpt-4-0613",
        "gpt-4-32k",
        "gpt-4-32k-0314",
        "gpt-4-32k-0613",
        "gpt-4-turbo",
        "gpt-4-turbo-2024-04-09",
        "gpt-4-1106-preview",
        "gpt-4-0125-preview",
        "gpt-4-vision-preview",
        "gpt-4-1106-vision-preview",
        "gpt-3.5-turbo",
        "gpt-3.5-turbo-0301",
        "gpt-3.5-turbo-0613",
        "gpt-3.5-turbo-1106",
        "gpt-3.5-turbo-0125",
        "gpt-3.5-turbo-16k",
        "gpt-3.5-turbo-16k-0613",
        "ft:gpt-3.5-turbo",
        "ft:gpt-3.5-turbo-0125",
        "ft:gpt-3.5-turbo-1106",
        "ft:gpt-3.5-turbo-0613",
        "ft:gpt-4-0613",
        "ft:gpt-4o-2024-08-06",
        "ft:gpt-4o-mini-2024-07-18",
        "text-embedding-3-large",
        "text-embedding-3-small",
        "text-embedding-ada-002",
        "text-embedding-ada-002-v2",
        "text-moderation-stable",
        "text-moderation-007",
        "text-moderation-latest",
        "256-x-256/dall-e-2",
        "512-x-512/dall-e-2",
        "1024-x-1024/dall-e-2",
        "hd/1024-x-1792/dall-e-3",
        "hd/1792-x-1024/dall-e-3",
        "hd/1024-x-1024/dall-e-3",
        "standard/1024-x-1792/dall-e-3",
        "standard/1792-x-1024/dall-e-3",
        "standard/1024-x-1024/dall-e-3",
        "whisper-1",
        "tts-1",
        "tts-1-hd"
    ],
    "message": null
}
```

Model Params API
```sh
curl --location 'http://127.0.0.1:4000/raga/internal/providers/models/params' \
--header 'Content-Type: application/json' \
--data '{
    "provider": "openai",
    "model": "gpt-4"
}'
```
Response:
```json
{
    "success": true,
    "data": [
        "frequency_penalty",
        "logit_bias",
        "logprobs",
        "top_logprobs",
        "max_tokens",
        "max_completion_tokens",
        "modalities",
        "n",
        "presence_penalty",
        "seed",
        "stop",
        "stream",
        "stream_options",
        "temperature",
        "top_p",
        "tools",
        "tool_choice",
        "function_call",
        "functions",
        "max_retries",
        "extra_headers",
        "parallel_tool_calls",
        "response_format"
    ],
    "message": null
}
```

Model Param values API:
```sh
curl --location 'http://localhost:4000/raga/internal/providers/models/params/values' \
--header 'Content-Type: application/json' \
--data '{
    "provider": "groq",
    "model": "groq/llama2-70b-4096"
}'
```
Response:
```json
{
    "success": true,
    "data": {
        "paramsValues": {
            "max_tokens": 4096,
            "max_input_tokens": 4096,
            "max_output_tokens": 4096,
            "input_cost_per_token": 7e-07,
            "output_cost_per_token": 8e-07,
            "supports_function_calling": true
        }
    },
    "message": null
}
```

### CODE REVIEW

**Review by kiljag — APPROVED** [review]

File: litellm/llms/cohere/completion.py [hunk 1]
```diff
@@ -153,7 +153,12 @@ def completion(
     headers = validate_environment(api_key, headers=headers)
     completion_url = api_base
     model = model
-    prompt = " ".join(message["content"] for message in messages)
+    prompt = ""
+    for message in messages:
+        if isinstance(message["content"], str):
+            prompt = prompt + " " + message["content"]
+        else:
+            prompt = prompt + "\n" + " ".join(item.get("text") for item in message["content"])
```
Comment (vibhanshu-ob) [inline litellm/llms/cohere/completion.py]: Due to change in structure
Previous:
```json
{
  "messages": [
    {
      "role": "system",
      "content": "What is automation testing ",
      "name": null,
      "function_call": null
    }
  ]
}
```
Current :
```json
{
  "messages": [
    {
      "role": "system",
      "content": [
        {
          "type": "text",
          "text": "What is automation testing "
        }
      ],
      "name": null,
      "function_call": null
    }
  ]
}
```

Comment (kiljag) [reply litellm/llms/cohere/completion.py]: This is not the right place to do it.

Comment (kiljag) [reply litellm/llms/cohere/completion.py]: We should be doing this in raga_utils

File: litellm/proxy/proxy_server.py [hunk 2]
```diff
@@ -3202,7 +3203,7 @@ async def chat_completion(  # noqa: PLR0915
         data, err = modify_user_request(data)
         if err is not None:
             return err
-
+        print(f"final payload: {json.dumps(data)}")
```
Comment (kiljag) [inline litellm/proxy/proxy_server.py]: This has secrets at this point, remove this print

Comment (vibhanshu-ob) [reply litellm/proxy/proxy_server.py]: removed

File: litellm/llms/cohere/completion.py (hunk 1 above)
Comment (vibhanshu-ob) [reply litellm/llms/cohere/completion.py]: removed

File: litellm/proxy/raga/raga_utils.py [hunk 3]
```diff
@@ -1,72 +1,59 @@
...
+        return None, {"error": {"message": f"Error: {str(e)}", "code": 400}}
 
 
 def set_api_keys_from_vault(data):
     print(f"getting api keys for user: {data['user_id']}")
     import litellm.proxy.raga.vault as vault
-    vault_secrets = vault.get_api_keys(data['user_id'])
 
-    model_name = data["model"]
-    if model_name.startswith("gpt"):
```
Comment (kiljag) [inline litellm/proxy/raga/raga_utils.py:41]: why are we deleting this check?

Comment (vibhanshu-ob) [reply litellm/proxy/raga/raga_utils.py:41]: It's handled in else block

**Review by kiljag — APPROVED** [review]

**Pull request merged**